cdap-ui/app/features/tracker/services/my-lineage-service.js

```diff
@@ -0,0 +1,226 @@
+
+        nodes.push({
+          dataId: dataId,
+          uniqueNodeId: rel.data,
+          isRightEdge: rel.data !== currentActiveNode
+        });
+        nodes.push({
+          dataId: programId,
+          uniqueNodeId: rel.program
+        });
+      }
+
+      uniqueNodes[rel.program].runs = uniqueNodes[rel.program].runs.concat(rel.runs);
+      uniqueNodes[rel.program].runs = _.uniq(uniqueNodes[rel.program].runs);
+    });
+
+    nodes = _.uniq(nodes, (n) => { return n.dataId; });
+    let graph = this.getGraphLayout(nodes, connections);
+    this.mapNodesLocation(nodes, graph);
+
+    return {
+      connections: connections,
+      nodes: nodes,
+      uniqueNodes: uniqueNodes,
+      graph: graph
+    };
+  }
+
+  parseProgramType(programType) {
+    let program = '';
+    switch (programType) {
+      case 'Flow':
+        program = 'flows';
+        break;
+      case 'Mapreduce':
+        program = 'mapreduce';
+        break;
+      case 'Spark':
+        program = 'spark';
+        break;
+      case 'Worker':
+        program = 'workers';
+        break;
+      case 'Workflow':
+        program = 'workflows';
+        break;
+      case 'Service':
+        program = 'services';
+        break;
+    }
+
+    return program;
+  }
+
+  parseDataInfo(data) {
+    let obj = {};
+    if (data.entityId.type === 'datasetinstance') {
+      obj = {
+        name: data.entityId.id.instanceId,
+        type: 'datasets',
+        icon: 'icon-datasets',
+        displayType: 'Dataset'
+      };
+    } else {
+      obj = {
+        name: data.entityId.id.streamName,
+        type: 'streams',
+        icon: 'icon-streams',
+        displayType: 'Stream'
+      };
+    }
+
+    return obj;
+  }
+
+  getProgramIcon(programType) {
+    let iconMap = {
+      'Flow': 'icon-tigon',
+      'Mapreduce': 'icon-mapreduce',
+      'Spark': 'icon-spark',
+      'Worker': 'icon-worker',
+      'Workflow': 'icon-workflow',
+      'Service': 'icon-service'
+    };
+
+    return iconMap[programType];
+  }
+
+  getGraphLayout(nodes, connections) {
+    var graph = new dagre.graphlib.Graph();
+    graph.setGraph({
+      nodesep: 90,
+      ranksep: 50,
+      rankdir: 'LR',
+      marginx: 90,
+      marginy: 25
+    });
+    graph.setDefaultEdgeLabel(function() { return {}; });
+
+    angular.forEach(nodes, (node) => {
+      var id = node.dataId;
+      graph.setNode(id, { width: 175, height: 55 });
+    });
+
+    angular.forEach(connections, (connection) => {
+      graph.setEdge(connection.source, connection.target);
+    });
+
+    dagre.layout(graph);
+
+    return graph;
+  }
+
+  mapNodesLocation(nodes, graph) {
+    angular.forEach(nodes, (node) => {
+      node._uiLocation = {
+        top: graph._nodes[node.dataId].y + 'px',
+        left: graph._nodes[node.dataId].x + 'px'
+      };
+    });
+  }
+}
+
+myLineageService.$inject = [];
+
+angular.module(PKG.name + '.feature.tracker')
+  .service('myLineageService', myLineageService);
```

Comment:
more like convention, anything we prefix with 'my' start with lowercase. The flux services is UpperCase